com.unity.shadergraph/Editor/Drawing/Blackboard/SGBlackboardCategory.cs

```diff
@@ -0,0 +1,623 @@
+using System;
+using System.Collections.Generic;
+using UnityEditor.Experimental.GraphView;
+using UnityEditor.ShaderGraph.Drawing.Views;
+using UnityEditor.ShaderGraph.Internal;
+using UnityEngine;
+using UnityEngine.UIElements;
+using System.Linq;
+
+using ContextualMenuManipulator = UnityEngine.UIElements.ContextualMenuManipulator;
+
+namespace UnityEditor.ShaderGraph.Drawing
+{
+    sealed class SGBlackboardCategory : GraphElement, ISGControlledElement<BlackboardCategoryController>, ISelection
+    {
+        // --- Begin ISGControlledElement implementation
+        public void OnControllerChanged(ref SGControllerChangedEvent e)
+        {
+        }
+
+        public void OnControllerEvent(SGControllerEvent e)
+        {
+        }
+
+        public BlackboardCategoryController controller
+        {
+            get => m_Controller;
+            set
+            {
+                if (m_Controller != value)
+                {
+                    if (m_Controller != null)
+                    {
+                        m_Controller.UnregisterHandler(this);
+                    }
+                    m_Controller = value;
+
+                    if (m_Controller != null)
+                    {
+                        m_Controller.RegisterHandler(this);
+                    }
+                }
+            }
+        }
+
+        SGController ISGControlledElement.controller => m_Controller;
+
+        // --- ISGControlledElement implementation
+
+        BlackboardCategoryController m_Controller;
+
+        BlackboardCategoryViewModel m_ViewModel;
+        public BlackboardCategoryViewModel viewModel => m_ViewModel;
+
+        const string k_StylePath = "Styles/SGBlackboard";
+        const string k_UxmlPath = "UXML/Blackboard/SGBlackboardCategory";
+
+        VisualElement m_DragIndicator;
+        VisualElement m_MainContainer;
+        VisualElement m_Header;
+        Label m_TitleLabel;
+        Foldout m_Foldout;
+        TextField m_TextField;
+        internal TextField textField => m_TextField;
+        VisualElement m_RowsContainer;
+        int m_InsertIndex;
+        SGBlackboard blackboard => m_ViewModel.parentView as SGBlackboard;
+
+        public delegate bool CanAcceptDropDelegate(ISelectable selected);
+
+        public CanAcceptDropDelegate canAcceptDrop { get; set; }
+
+        int InsertionIndex(Vector2 pos)
```

Comment:
I've seen this logic in a few places, would it be possible to extract to some shared location?

Reply:
Extracted the core of the InsertionIndex function into `BlackboardUtils.GetInsertionIndex` that both `SGBlackboard` and `SGBlackboardCategory` call now 👍